src/ByteCountingStream.php

```diff
@@ -0,0 +1,64 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\StreamInterface;
+use SebastianBergmann\GlobalState\RuntimeException;
+
+/**
+ *  Stream decorator that counts the number of bytes read off of an underlying
+ *  read stream.
+ */
+class ByteCountingStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    /** @var int Number of bytes allowed to read */
+    private $maximum;
```

Comment:
Perhaps you can just tracking the `$remainingBytes` instead of tracking `$maximum` and `$bytesCount`?